### PR TITLE
Add functionality for moving blobs to 'rejected' containers

### PR DESCRIPTION
**Pull request by lgonczar**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RPE-679

### Change description ###

Add functionality for moving blobs to 'rejected' containers. This functionality is not used, yet (it will be in the following PR).

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java [hunk 1]
```diff
@@ -74,4 +81,66 @@ public CloudBlobContainer getContainer(String containerName) throws URISyntaxExc
...
+                fileName,
+                inputContainerName,
+                getRejectedContainerName(inputContainerName)
+            );
+        } catch (Exception ex) {
+            log.error(
+                "An error occurred when moving rejected file {} from container {} to rejected files' container {}",
+                fileName,
+                inputContainerName,
+                getRejectedContainerName(inputContainerName),
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java]: What do you mean?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java [hunk 2]
```diff
@@ -74,4 +81,66 @@ public CloudBlobContainer getContainer(String containerName) throws URISyntaxExc
...
+    private void waitUntilBlobIsCopied(CloudBlockBlob blob)
+        throws InterruptedException, StorageException {
+
+        CopyStatus copyStatus = CopyStatus.PENDING;
+        boolean timeout = false;
+        DateTime startTime = DateTime.now();
+
+        do {
+            if (DateTime.now().minusMillis(properties.getBlobCopyTimeoutInMillis()).isAfter(startTime)) {
+                timeout = true;
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java]: We need to set (and handle) some timeout for copying the file, anyway.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java [hunk 3]
```diff
@@ -74,4 +81,66 @@ public CloudBlobContainer getContainer(String containerName) throws URISyntaxExc
...
+            } else {
+                Thread.sleep(properties.getBlobCopyPollingDelayInMillis());
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java]: No, it defaults to 500 ms.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java]: extract repetitive call to `String rejectedContainerName = getRejectedContainerName(inputContainerName);` and then do try/catch

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java [hunk 4]
```diff
@@ -74,4 +81,66 @@ public CloudBlobContainer getContainer(String containerName) throws URISyntaxExc
...
+        } while (copyStatus == CopyStatus.PENDING && !timeout);
+
+        if (copyStatus != CopyStatus.SUCCESS) {
+            String errorMessage = timeout
+                ? "Timed out while waiting for rejected file to be copied"
+                : String.format("Unexpected status of copy operation: %s", copyStatus);
+
+            throw new RejectedBlobCopyException(errorMessage);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/BlobManager.java]: I don't think this is the best use case for DelayQueue. The solution is more complicated than it needs to be.

**Pull request merged**